### PR TITLE
Move to hatch based setup. Lint with Ruff. Add mypy.

### PR DESCRIPTION
**Pull request by BSchilperoort**
- Moved to hatch
- Set up hatch envs and scripts
- Formatted all code with black
- Changed linting to ruff
  - Mostly the same as previous, with added Pathlib rules
- Became stricter about Path usage.
- Added mypy
  - Required changes to type hinting (and removing the types from docstrings).

### CODE REVIEW

**Review by geek-yang — APPROVED** [review]
Well done! Looks quite neat! I only have some very small comments. Feel free to address them and merge this PR 😄.

File: PyStemmusScope/utils.py [hunk 1]
```diff
@@ -85,36 +89,50 @@ def to_absolute_path(
...
     return pathlike.expanduser().resolve(strict=must_exist)
 
 
 def get_forcing_file(config):
-    """Get forcing file from the location.
-    """
+    """Get forcing file from the location."""
     location, fmt = check_location_fmt(config["Location"])
-    # check if the forcing file exists for the given locaiton(s)
+    # check if the forcing file exists for the given locations(s)
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply PyStemmusScope/utils.py:100]: Oops yes. You're right :)

File: PyStemmusScope/utils.py [hunk 2]
```diff
@@ -85,36 +89,50 @@ def to_absolute_path(
...
+    else:
+        raise ValueError("Unknown value for site format")
 
     return forcing_file
 
 
-def check_location_fmt(loc):
+# Location format aliases
+LatLonFmt = Tuple[float, float]
+BBoxFmt = Tuple[LatLonFmt, LatLonFmt]
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply PyStemmusScope/utils.py:132]: I put them next to this function so I could actually make the type definition readable. I generally find that once the type alias is moved to far away it becomes less useful.

There is of course also PEP-20 which says `Readability counts.` And of course "[A Foolish Consistency is the Hobgoblin of Little Minds](https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds)". 😉

File: global_data/download_scripts/download_era5land_monthly.py [hunk 3]
```diff
@@ -32,32 +30,71 @@
...
+                "22",
+                "23",
+                "24",
+                "25",
+                "26",
+                "27",
+                "28",
+                "29",
+                "30",
                 "31",
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply global_data/download_scripts/download_era5land_monthly.py:63]: Oh I think black formatted it like this, and I missed that it had updated this... I'll fix it

File: pyproject.toml [hunk 4]
```diff
@@ -1,21 +1,179 @@
...
+
+[project]
+name = "PyStemmusScope"
+description = "Python modules for running the STEMMUS-SCOPE model."
+readme = "README.md"
+license = "Apache-2.0"
+requires-python = ">=3.8, <3.11"
+authors = [
+  {email = "f.alidoost@esciencecenter.nl"},
+  {name = "Sarah Alidoost"}
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply pyproject.toml]: I wish PyPI was better at parsing multiple authors from the pyproject.toml file, I believe it only takes a single one, despite the field being called `authors`

**Pull request merged**